trestle/core/catalog/catalog_reader.py

```diff
@@ -152,3 +223,146 @@ def _read_comp_info_from_md(control_file_path: pathlib.Path,
+        param_values = [com.Value(__root__=param_val) for param_val in param_dict['values']]
+        for comp_rule_param in comp_rules_params:
+            if comp_rule_param['name'] == param_name:
+                rule_id = comp_rule_param[const.HEADER_RULE_ID]
+                CatalogReader._insert_set_param_into_by_comps(imp_req, rule_id, param_name, param_values)
+                for statement in as_list(imp_req.statements):
+                    CatalogReader._insert_set_param_into_by_comps(statement, rule_id, param_name, param_values)
+
+    @staticmethod
+    def _update_ssp_with_md_header(
+        ssp: ossp.SystemSecurityPlan,
+        control_id: str,
+        comp_dict: Dict[str, generic.GenericComponent],
+        md_header: Dict[str, Dict[str, str]]
+    ) -> None:
+        # rules param vals go in bycomps of imp_req
+        # param vals go directly in imp_req
+        rules_param_vals_dict = md_header.get(const.COMP_DEF_RULES_PARAM_VALS_TAG, {})
+        imp_req = CatalogReader._get_imp_req_for_control(ssp, control_id)
+        for comp_name, param_dict_list in rules_param_vals_dict.items():
+            for param_dict in as_list(param_dict_list):
+                if const.SSP_VALUES in param_dict:
+                    # needs to go in statements that support the rule
+                    CatalogReader._insert_param_dict_in_imp_req(
+                        imp_req, param_dict, comp_name, comp_dict[comp_name].uuid, md_header
+                    )
+                    by_comp = CatalogReader._get_by_comp_from_imp_req(imp_req, '', comp_dict[comp_name].uuid)
+                    by_comp.set_parameters = as_list(by_comp.set_parameters)
+                    param_id = param_dict['name']
+                    value_list = param_dict[const.SSP_VALUES]
+                    param_values = [com.Value(__root__=value) for value in value_list]
+                    # remove any preceding set params for this param_id
+                    new_sp_list = []
+                    for sp in as_list(by_comp.set_parameters):
+                        if sp.param_id != param_id:
+                            new_sp_list.append(sp)
+                    by_comp.set_parameters = new_sp_list
+                    by_comp.set_parameters.append(ossp.SetParameter(param_id=param_id, values=param_values))
+        param_vals_dict = md_header.get(const.SET_PARAMS_TAG, {})
+        for param_id, param_dict in param_vals_dict.items():
+            if const.SSP_VALUES in param_dict:
+                value_list = param_dict[const.SSP_VALUES]
+                param_values = [com.Value(__root__=value) for value in value_list]
+                new_sp_list = []
+                for sp in as_list(imp_req.set_parameters):
+                    if sp.param_id != param_id:
+                        new_sp_list.append(sp)
+                imp_req.set_parameters = new_sp_list
+                imp_req.set_parameters.append(ossp.SetParameter(param_id=param_id, values=param_values))
+
```

Comment:
These lines are complete duplicate of lines 309-317 and a good example why separating function to the logical pieces is a good practice. Long specific functions are hard to reuse and debug.

Reply:
They are different in that they are operating on entirely different things: by_comp vs. imp_req.  But I split off a function that operates on a new TypeWithSetParams to remove the duplication.

Reply:
Both of them perform the same exact logic and operate on the List type.
Lines 309-317 and lines 321-328. Pass `by_comp.set_parameters` and `imp_req.set_parameters` to a function and it will avoid the duplication

Reply:
Both of them perform the same exact logic and operate on the List type.
Lines 309-317 and lines 321-328. Pass by_comp.set_parameters and imp_req.set_parameters to a function and it will avoid the duplication